File: sdk/runanywhere-react-native/packages/core/src/Infrastructure/FileManagement/Models/DeviceStorageInfo.ts
/**
 * DeviceStorageInfo.ts
 * RunAnywhere SDK
 *
 * Device storage information
 *
 * Reference: sdk/runanywhere-swift/Sources/RunAnywhere/Infrastructure/FileManagement/Models/Domain/DeviceStorageInfo.swift
 */

/**
 * Device storage information
 */
export interface DeviceStorageInfo {
  /** Total device storage space in bytes */
  readonly totalSpace: number;
  /** Free device storage space in bytes */
  readonly freeSpace: number;
  /** Used device storage space in bytes */
  readonly usedSpace: number;
}

/**
 * Creates a DeviceStorageInfo instance
 *
 * @param params - Device storage parameters
 * @returns DeviceStorageInfo instance
 */
export function createDeviceStorageInfo(params: {
  totalSpace: number;
  freeSpace: number;
  usedSpace: number;
}): DeviceStorageInfo {
  return {
    totalSpace: params.totalSpace,
    freeSpace: params.freeSpace,
    usedSpace: params.usedSpace,
  };
}

/**
 * Calculate device storage usage percentage
 *
 * @param info - DeviceStorageInfo instance
 * @returns Usage percentage (0-100)
 */
export function getDeviceStorageUsagePercentage(
  info: DeviceStorageInfo
): number {
  if (info.totalSpace <= 0) {
    return 0;
  }
  return (info.usedSpace / info.totalSpace) * 100;
}

File: sdk/runanywhere-flutter/packages/runanywhere/lib/core/models/storage/device_storage_info.dart
/// Device storage information.
/// Matches iOS DeviceStorageInfo from Data/Models/Storage/DeviceStorageInfo.swift
class DeviceStorageInfo {
  /// Total device storage space in bytes
  final int totalSpace;

  /// Free device storage space in bytes
  final int freeSpace;

  /// Used device storage space in bytes
  final int usedSpace;

  const DeviceStorageInfo({
    required this.totalSpace,
    required this.freeSpace,
    required this.usedSpace,
  });

  /// Usage percentage (0-100)
  double get usagePercentage {
    if (totalSpace == 0) return 0;
    return (usedSpace / totalSpace) * 100;
  }

  /// Create from JSON map
  factory DeviceStorageInfo.fromJson(Map<String, dynamic> json) {
    return DeviceStorageInfo(
      totalSpace: (json['totalSpace'] as num?)?.toInt() ?? 0,
      freeSpace: (json['freeSpace'] as num?)?.toInt() ?? 0,
      usedSpace: (json['usedSpace'] as num?)?.toInt() ?? 0,
    );
  }

  /// Convert to JSON map
  Map<String, dynamic> toJson() {
    return {
      'totalSpace': totalSpace,
      'freeSpace': freeSpace,
      'usedSpace': usedSpace,
    };
  }
}
